commit 0f6da15c3d8c233a89c3435663c49b506bf7a2e1
Author: Vadim Zeitlin <vadim@wxwidgets.org>
Date:   Fri Jan 24 18:11:36 2014 +0100

    Support 256*256 sizes icons in wxICOHandler.
    
    The size of such icons is 0 in the header (as it's one byte only), handle this
    special case correctly.

diff --git a/src/common/imagbmp.cpp b/src/common/imagbmp.cpp
index 20cdf50..060b46f 100644
--- a/src/common/imagbmp.cpp
+++ b/src/common/imagbmp.cpp
@@ -1161,8 +1161,8 @@ bool wxBMPHandler::DoCanRead(wxInputStream& stream)
 
 struct ICONDIRENTRY
 {
-    wxUint8         bWidth;               // Width of the image
-    wxUint8         bHeight;              // Height of the image (times 2)
+    wxUint8         bWidth;               // Width of the image (256 if 0)
+    wxUint8         bHeight;              // Height of the image (256 if 0)
     wxUint8         bColorCount;          // Number of colors in image (0 if >=8bpp)
     wxUint8         bReserved;            // Reserved
 
@@ -1189,7 +1189,7 @@ bool wxICOHandler::SaveFile(wxImage *image,
 
 {
     //sanity check; icon must be less than 127 pixels high and 255 wide
-    if ( image->GetHeight () > 127 )
+    if ( image->GetHeight () > 128 )
     {
         if ( verbose )
         {
@@ -1197,7 +1197,7 @@ bool wxICOHandler::SaveFile(wxImage *image,
         }
         return false;
     }
-    if ( image->GetWidth () > 255 )
+    if ( image->GetWidth () > 256 )
     {
         if ( verbose )
         {
@@ -1321,7 +1321,7 @@ bool wxICOHandler::SaveFile(wxImage *image,
 
         offset = offset + sizeof(ICONDIRENTRY);
 
-        icondirentry.bWidth = (wxUint8)image->GetWidth();
+        icondirentry.bWidth = (wxUint8)image->GetWidth(); // OK for 256
         icondirentry.bHeight = (wxUint8)(2 * image->GetHeight());
         icondirentry.bColorCount = 0;
         icondirentry.bReserved = 0;
@@ -1403,10 +1403,9 @@ bool wxICOHandler::LoadFile(wxImage *image, wxInputStream& stream,
 }
 
 bool wxICOHandler::DoLoadFile(wxImage *image, wxInputStream& stream,
-                            bool WXUNUSED(verbose), int index)
+                            bool verbose, int index)
 {
     bool bResult wxDUMMY_INITIALIZE(false);
-    bool IsBmp = false;
 
     ICONDIR IconDir;
 
@@ -1435,8 +1434,11 @@ bool wxICOHandler::DoLoadFile(wxImage *image, wxInputStream& stream,
 
         alreadySeeked += stream.LastRead();
 
+        const wxUint16
+            widthReal = pCurrentEntry->bWidth ? pCurrentEntry->bWidth : 256;
+
         // bHeight and bColorCount are wxUint8
-        if ( pCurrentEntry->bWidth >= wMax )
+        if ( widthReal >= wMax )
         {
             // see if we have more colors, ==0 indicates > 8bpp:
             if ( pCurrentEntry->bColorCount == 0 )
@@ -1444,7 +1446,7 @@ bool wxICOHandler::DoLoadFile(wxImage *image, wxInputStream& stream,
             if ( pCurrentEntry->bColorCount >= colmax )
             {
                 iSel = i;
-                wMax = pCurrentEntry->bWidth;
+                wMax = widthReal;
                 colmax = pCurrentEntry->bColorCount;
             }
         }
@@ -1475,7 +1477,7 @@ bool wxICOHandler::DoLoadFile(wxImage *image, wxInputStream& stream,
         if (offset != 0 && stream.SeekI(offset, wxFromCurrent) == wxInvalidOffset)
             return false;
 
-        bResult = LoadDib(image, stream, true, IsBmp);
+        bResult = LoadDib(image, stream, verbose, false /* not BMP */);
         bool bIsCursorType = (this->GetType() == wxBITMAP_TYPE_CUR) || (this->GetType() == wxBITMAP_TYPE_ANI);
         if ( bResult && bIsCursorType && nType == 2 )
         {
